README
==============================================
SOAPpy - Simple to use SOAP library for Python 
==============================================

Current Maintainer: 	

	jeff kit <jeffkit.info@gmail.com>

Original Authors: 

	Cayce Ullman <c_ullman@yahoo.com>
        Brian Matthews <blm@blmatthews.com>

Contributions by:

	Christopher Blunck <blunck2@gst.com>
	Brad Knotwell <b.knotwell@f5.com>
	Mark Bucciarelli <mark@hubcapconsulting.com> (ported WSDL
             client from ZSI)
	Ivan R. Judson 	<judson@mcs.anl.gov> (Globus support)
	Kirk Strauser <kirk@daycos.com>
	Antonio Beamud Montero <antonio.beamud@linkend.com> (patches
  	     for integrating SOAPpy into Zope)
	And others.

Copyright (c) 2002-2005, Pfizer, Inc.
Copyright (c) 2001, Cayce Ullman.
Copyright (c) 2001, Brian Matthews.
All rights reserved, see the file LICENSE for conditions of use.

INTRODUCTION
============

    The goal of the SOAPpy team is to provide a full-featured SOAP library
    for Python that is very simple to use and that fully supports dynamic
    interaction between clients and servers.
    
    INCLUDED
    --------

    - General SOAP Parser based on sax.xml
    - General SOAP Builder
    - SOAP Proxy for RPC client code
    - SOAP Server framework for RPC server code
    
    FEATURES
    --------

    - Handles all SOAP 1.0 types
    - Handles faults 
    - Allows namespace specification
    - Allows SOAPAction specification
    - Homogeneous typed arrays
    - Supports multiple schemas
    - Header support (mustUnderstand and actor)
    - XML attribute support
    - Multi-referencing support (Parser/Builder)
    - Understands SOAP-ENC:root attribute
    - Good interop, passes all client tests for Frontier, SOAP::LITE, SOAPRMI
    - Encodings
    - SSL clients (with Python compiled with OpenSSL support)
    - SSL servers (with Python compiled with OpenSSL support and M2Crypto
      installed)
    - Encodes XML tags per SOAP 1.2 name mangling specification (Gregory Warnes)
    - Automatic stateful SOAP server support (Apache v2.x) (blunck2)
    - WSDL client support
    - WSDL server support
    
    TODO (See RELEASE_INFO and CHANGELOG for recent changes)
    ----

    - Timeout on method calls
    - Advanced arrays (sparse, multidimensional and partial)
    - Attachments
    - mod_python example
    - medusa example
    - Improved documentation
    
    MANIFEST
    --------
    
    Files

    
        README	        This file
        RELEASE_NOTES       General information about each release
        ChangeLog           Detailed list of changes
        TODO		List of tasks that need to be done
    
        setup.py            Python installation control files
        MANIFEST
        MANIFEST.in
        
        SOAPpy.spec*        RPM package control file
    
    Directories
    
        SOAPpy/*            Source code for the package
        SOAPpy/wstools/*    Source code for WSDL tools
        tests/*             unit tests and examples
        validate/*          interop client and servers
        bid/*		N+I interop client and server
        doc/*		Documentation
        contrib/		Contributed examples (also see test/)
        docs/		Documentation
        tools/              Misc tools useful for the SOAPpy developers
        zope/               Patches to Zope allowing it to provide SOAP services
    

INSTALLATION
============

    REQUIRED PACKAGES:
    -----------------

    - fpconst 0.6.0 or later,
      <http://research.warnes.net/projects/rzope/fpconst/>

    - pyXML 0.8.3 or later, <http://pyxml.sourceforge.net>
    
    OPTIONAL PACKAGES
    -----------------
    
    - pyGlobus, optional support for Globus,
      <http://www-itg.lbl.gov/gtg/projects/pyGlobus/>
    
    - M2Crypto.SSL, optional support for server-side SSL
      <http://sandbox.rulemaker.net/ngps/m2/>
    
    - If Python is compiled with SSL support (Python 2.3 does so by
      default), client-side use of SSL is supported
    
    INSTALLATION STEPS
    ------------------
    
    As of version 0.9.8 SOAPpy can be installed using the standard python
    package installation tools.  
    
    To install:
    
      1) Unpack the distribution package:
    
         On Windows, use your favorite zip file uncompression tool.
    
         On Unix: 
        
             $ tar -xvzf SOAPpy-$VERSION$.tar.gz
        
             if you have gnu tar, otherwise
        
             $ gzcat SOAPpy-$VERSION$.tar.gz | tar -xvf -
        
      2) Change into the source directory
    
         $ cd SOAPpy-$VERSION$
    
      3) Compile the package 
    
         $ python setup.py build
    
      4) Install the package

	 On Windows:

	     $ python setup.py install

	 On Unix install as the owner of the python directories
	 (usally root):
    
	     $ su root
             Password: XXXXXX
             $ python setup.py install
    
    
DOCUMENTATION
=============

    QUICK START
    -----------

    A simple "Hello World" http SOAP server:

        import SOAPpy
        def hello():
	    return "Hello World"

        server = SOAPpy.SOAPServer(("localhost", 8080))
        server.registerFunction(hello)
        server.serve_forever()

    And the corresponding client:

        import SOAPpy
        server = SOAPpy.SOAPProxy("http://localhost:8080/")
        print server.hello()

    BASIC TUTORIAL
    --------------

    Mark Pilgrims' _Dive Into Python_, published in printed form by
    Apress and online at at http://diveintopython.org provides a
    nice tutorial for SOAPpy in Chapter 12, "SOAP Web Services".
    See http://diveintopython.org/soap_web_services .

    OTHER DOCUMENTATION
    -------------------
      
    For further information see the files in the docs/ directory.

    Note that documentation is one of SOAPpy's current weak points.
    Please help us out!


GETTING HELP
============

    REPORTING BUGS
    --------------
    
    Please submit bug reports, feature requests, patches, etc at the
    Python Web Services web site:  http://pywebsvcs.sourceforge.net.
    
    MAILING LIST
    ============
    
    Please address questions and general discussion to the
    pywebsvcs-talk mailing list, pywebsvcs-talk@lists.sourceforge.net.
    
    For subscription information visit
    http://lists.sourceforge.net/lists/listinfo/pywebsvcs-talk.  
    List archives are available at
    http://sourceforge.net/mailarchive/forum.php?forum=pywebsvcs-talk

    Please remember that the authors do have day jobs, so please try
    the mailing list before contacting them directy.  
        
$Id: README,v 1.18 2005/02/22 15:58:35 warnes Exp $


ChangeLog
2011-03-05 22:00 jeff
    
    * Client.py, Types.py, GSIServer.py,NS.py,Server.py,SOAPBuilder.py,README,ChangeLog： update for 0.12.1 release.

2005-02-22 10:58  warnes

	* ChangeLog, README, RELEASE_INFO: Update for 0.12.0 release.

2005-02-21 23:30  tag SOAPpy_0_12_0

2005-02-21 23:30  warnes

	* tests/README: update README since TCtests.py now passes!

2005-02-21 23:29  warnes

	* SOAPpy/: Parser.py, Types.py: Fix handling of parse rules!
	  TCtest.py now passes!

2005-02-21 23:25  warnes

	* tests/: Bug1001646.py, Bug916265.py, ComplexTypes.py,
	  GoogleTest.py, Makefile, README, TCtest.py, echoClient.py,
	  echoHeader.py, echoServer.py, largeDataTest.py, testleak.py:
	  Enhance echoServer with a quit method.  Provide a makefile for
	  running all working tests.

2005-02-21 15:27  warnes

	* SOAPpy/Client.py: Fix loss of quotes at start of  file
	  description.

2005-02-21 15:24  warnes

	* SOAPpy/SOAPBuilder.py: Add spaces and newlines to soap
	  environment declarations when writing SOAP enclosure so that the
	  results are more human readable.

2005-02-21 15:21  warnes

	* SOAPpy/Client.py: Add code to handle the case when headers are
	  empty or not present.

2005-02-21 15:16  warnes

	* SOAPpy/WSDL.py: Add 'show_methods' which will print the methods
	  and associated parameters.

2005-02-21 15:09  warnes

	* docs/: GettingStarted.txt, GlobusSupport.txt, WSDL.txt,
	  complexTypes.txt, simpleTypes.txt: Update documentation.

2005-02-18 14:29  warnes

	* SOAPpy/Client.py: If content_length is missing or empty, we need
	  not check for a the OC4J bug...

2005-02-18 11:28  warnes

	* README: Add bug reporting and mailing list information.

2005-02-18 10:42  warnes

	* docs/GettingStarted.txt: Change 'SOAP.py' to 'SOAPpy'

2005-02-18 10:40  warnes

	* docs/attrs.txt: Remove the long copyright and usage notices.
	  Clarify the (too brief) text.

2005-02-18 10:36  warnes

	* docs/: complexTypes.txt, UsingHeaders.txt: Most of the
	  information in the 'complexTypes.txt' file was actually about the
	  use of headers.  Moved that text to a new file
	  'UsingHeaders.txt'.

2005-02-18 08:50  warnes

	* SOAPpy/wstools/XMLSchema.py: Apply patch submitted by Peter
	  McA'Nulty of WebReply.com:	 "class SimpleContent, class
	  Extension.fromDom() at line ~2313	 doesn't handle extensions
	  without contents -- unusual, but not	    illegal"

2005-02-17 01:44  boverhof

	* SOAPpy/wstools/XMLSchema.py:
	  ----------------------------------------------------------------------

	   Modified Files:
		  XMLSchema.py -- added the item trace to an exception
	  thrown
		     when an unexpected child is encountered in
	  complexType.fromDom
		     method so one can see where the problem is occuring.


	  ----------------------------------------------------------------------

2005-02-16 13:39  warnes

	* tests/echoServer.py: Added echo_jheader which returns its
	  argument and the header, in order to support Bug1001646.py

2005-02-16 13:34  warnes

	* tests/Bug1001646.py: Test now properly checks to see that the
	  header is preserved.

2005-02-16 09:45  tag v1_7

2005-02-16 09:45  warnes

	* SOAPpy/wstools/XMLname.py: Remember to paste prefix back onto the
	  XMLname before returning.

2005-02-15 23:27  warnes

	* SOAPpy/SOAPBuilder.py: Bug fix for [ 1078051 ] "Arrays of complex
	  types (doc/lit)" per the patch submited with the bug report by
	  Nelson Minar.

2005-02-15 23:25  warnes

	* tests/echoClient.py: Print what we're trying to do so that errors
	  are easier to track down.

2005-02-15 23:24  warnes

	* tests/storageTest.py: Updated 'registerUser' to match current
	  documentation, but still get an error from the server.

2005-02-15 23:23  warnes

	* tests/echoServer.py: Add 'echo_simple' command that just returns
	  its arguments as-is.

2005-02-15 23:22  warnes

	* tests/Bug1001646.py: Turn of some debugging information.

2005-02-15 23:21  warnes

	* tests/Bug916265.py: Regression test for [ 916265 ] "Arrays of
	  unicode do not serialize correctly (patch included)"

2005-02-15 23:09  warnes

	* tests/Bug1001646.py: Regression test for bug [ 1001646 ] SOAPpy
	  stomps headers when sending multirefs.

2005-02-15 16:51  warnes

	* SOAPpy/Client.py: Create workaround for a bug in OC4J reported in
	  [ 1122991 ] "error from SOAPpy/Client.py for content_length
	  evaluation?" by Merten Schumann.

2005-02-15 16:09  warnes

	* SOAPpy/SOAPBuilder.py: Fix [ 1106450 ] "Floats are truncated to
	  10 digits, causing precision loss" submitted by Kerry 'krdavis'.

2005-02-15 16:07  warnes

	* SOAPpy/Client.py: Fixed [ 1096971 ] "Parse error: missing HTTP
	  header 'Content-length'" submitted by 'pure water'.

2005-02-15 15:59  warnes

	* SOAPpy/Types.py: Fix [ 1064248 ] "Bugs in _asdict() and
	  _asarray() in Types.py" submitted by Peter Lamb.

2005-02-15 15:56  warnes

	* SOAPpy/SOAPBuilder.py: Fix handling of 0-length arrays.

2005-02-15 15:52  warnes

	* SOAPpy/SOAPBuilder.py: Apply [ 1064233 ] "Bug fixes for complex
	  types" from Peter Lamb.

2005-02-15 15:41  warnes

	* SOAPpy/SOAPBuilder.py: Fix bug [ 1001646 ] SOAPpy stomps headers
	  when sending multirefs using included patch provide by Nelson
	  Minar

2005-02-15 15:15  warnes

	* SOAPpy/Client.py: Fix [ 925077 ] SOAPpy prints out SOAP fault
	  /even when Config.debug is off/.

2005-02-15 15:12  warnes

	* SOAPpy/Parser.py, tests/Bug918216.py: Fix parsing bug & add
	  regression test.  SOAPpy was not allowing anything after the
	  close of a faulttype block, but multirefs should be permitted
	  there.  Closes bug [ 918216 ] "Parsing faults in SOAPpy 0.11.3"

2005-02-15 14:30  warnes

	* tests/ZeroLengthArray.py: Test handling of zero-length typed
	  lists.  Currently fails.

2005-02-15 14:22  warnes

	* SOAPpy/SOAPBuilder.py: Revert broken header generation from last
	  patch.

2005-02-15 11:41  warnes

	* SOAPpy/SOAPBuilder.py: Fix bug [ 916265 ] Arrays of unicode do
	  not serialize correctly, submitted by Nelson Minar.

2005-02-15 11:37  warnes


tests/BabelfishWSDLTest.py
#!/usr/bin/env python

ident = '$Id: BabelfishWSDLTest.py,v 1.1 2003/07/18 15:58:28 warnes Exp $'

import os, re
import sys
sys.path.insert(1, "..")

from SOAPpy import WSDL

# Check for a web proxy definition in environment
try:
   proxy_url=os.environ['http_proxy']
   phost, pport = re.search('http://([^:]+):([0-9]+)', proxy_url).group(1,2)
   proxy = "%s:%s" % (phost, pport)
except:
   proxy = None

server = WSDL.Proxy('http://www.xmethods.net/sd/2001/BabelFishService.wsdl',
                    http_proxy=proxy)

english = "Hi Friend!"

print "Babelfish Translations"
print "------------------------"
print "English: '%s'" % english
print "French:  '%s'" % server.BabelFish('en_fr',english)
print "Spanish: '%s'" % server.BabelFish('en_es',english)
print "Italian: '%s'" % server.BabelFish('en_it',english)
print "German:  '%s'" % server.BabelFish('en_de',english)

print "Done."


SOAPpy/NS.py
"""
################################################################################
#
# SOAPpy - Cayce Ullman       (cayce@actzero.com)
#          Brian Matthews     (blm@actzero.com)
#          Gregory Warnes     (Gregory.R.Warnes@Pfizer.com)
#          Christopher Blunck (blunck@gst.com)
#
################################################################################
# Copyright (c) 2003, Pfizer
# Copyright (c) 2001, Cayce Ullman.
# Copyright (c) 2001, Brian Matthews.
#
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
# Redistributions of source code must retain the above copyright notice, this
# list of conditions and the following disclaimer.
#
# Redistributions in binary form must reproduce the above copyright notice,
# this list of conditions and the following disclaimer in the documentation
# and/or other materials provided with the distribution.
#
# Neither the name of actzero, inc. nor the names of its contributors may
# be used to endorse or promote products derived from this software without
# specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE FOR
# ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
################################################################################
"""
from __future__ import nested_scopes

ident = '$Id: NS.py,v 1.4 2005/02/15 16:32:22 warnes Exp $'
from version import __version__

##############################################################################
# Namespace Class
################################################################################
def invertDict(dict):
    d = {}

    for k, v in dict.items():
        d[v] = k

    return d

class NS:
    XML  = "http://www.w3.org/XML/1998/namespace"

    ENV  = "http://schemas.xmlsoap.org/soap/envelope/"
    ENC  = "http://schemas.xmlsoap.org/soap/encoding/"

    XSD  = "http://www.w3.org/1999/XMLSchema"
    XSD2 = "http://www.w3.org/2000/10/XMLSchema"
    XSD3 = "http://www.w3.org/2001/XMLSchema"

    XSD_L = [XSD, XSD2, XSD3]
    EXSD_L= [ENC, XSD, XSD2, XSD3]

    XSI   = "http://www.w3.org/1999/XMLSchema-instance"
    XSI2  = "http://www.w3.org/2000/10/XMLSchema-instance"
    XSI3  = "http://www.w3.org/2001/XMLSchema-instance"
    XSI_L = [XSI, XSI2, XSI3]

    URN   = "http://soapinterop.org/xsd"

    # For generated messages
    XML_T = "xml"
    ENV_T = "SOAP-ENV"
    ENC_T = "SOAP-ENC"
    XSD_T = "xsd"
    XSD2_T= "xsd2"
    XSD3_T= "xsd3"
    XSI_T = "xsi"
    XSI2_T= "xsi2"
    XSI3_T= "xsi3"
    URN_T = "urn"

    NSMAP       = {ENV_T: ENV, ENC_T: ENC, XSD_T: XSD, XSD2_T: XSD2,
                    XSD3_T: XSD3, XSI_T: XSI, XSI2_T: XSI2, XSI3_T: XSI3,
                    URN_T: URN}
    NSMAP_R     = invertDict(NSMAP)

    STMAP       = {'1999': (XSD_T, XSI_T), '2000': (XSD2_T, XSI2_T),
                    '2001': (XSD3_T, XSI3_T)}
    STMAP_R     = invertDict(STMAP)

    def __init__(self):
        raise Error, "Don't instantiate this"





SOAPpy/Client.py
"""
################################################################################
#
# SOAPpy - Cayce Ullman       (cayce@actzero.com)
#          Brian Matthews     (blm@actzero.com)
#          Gregory Warnes     (Gregory.R.Warnes@Pfizer.com)
#          Christopher Blunck (blunck@gst.com)
#
################################################################################
# Copyright (c) 2003, Pfizer
# Copyright (c) 2001, Cayce Ullman.
# Copyright (c) 2001, Brian Matthews.
#
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
# Redistributions of source code must retain the above copyright notice, this
# list of conditions and the following disclaimer.
#
# Redistributions in binary form must reproduce the above copyright notice,
# this list of conditions and the following disclaimer in the documentation
# and/or other materials provided with the distribution.
#
# Neither the name of actzero, inc. nor the names of its contributors may
# be used to endorse or promote products derived from this software without
# specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE FOR
# ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
################################################################################
"""
from __future__ import nested_scopes

ident = '$Id: Client.py,v 1.27 2005/02/21 20:27:09 warnes Exp $'
from version import __version__


#import xml.sax
import urllib
from types import *
import re
import base64

# SOAPpy modules
from Errors      import *
from Config      import Config
from Parser      import parseSOAPRPC
from SOAPBuilder import buildSOAP
from Utilities   import *
from Types       import faultType, simplify

################################################################################
# Client
################################################################################


def SOAPUserAgent():
    return "SOAPpy " + __version__ + " (pywebsvcs.sf.net)"


class SOAPAddress:
    def __init__(self, url, config = Config):
        proto, uri = urllib.splittype(url)

        # apply some defaults
        if uri[0:2] != '//':
            if proto != None:
                uri = proto + ':' + uri

            uri = '//' + uri
            proto = 'http'

        host, path = urllib.splithost(uri)

        try:
            int(host)
            host = 'localhost:' + host
        except:
            pass

        if not path:
            path = '/'

        if proto not in ('http', 'https', 'httpg'):
            raise IOError, "unsupported SOAP protocol"
        if proto == 'httpg' and not config.GSIclient:
            raise AttributeError, \
                  "GSI client not supported by this Python installation"
        if proto == 'https' and not config.SSLclient:
            raise AttributeError, \
                "SSL client not supported by this Python installation"

        self.user,host = urllib.splituser(host)
        self.proto = proto
        self.host = host
        self.path = path

    def __str__(self):
        return "%(proto)s://%(host)s%(path)s" % self.__dict__

    __repr__ = __str__


class HTTPTransport:
    def getNS(self, original_namespace, data):
        """Extract the (possibly extended) namespace from the returned
        SOAP message."""

        if type(original_namespace) == StringType:
            pattern="xmlns:\w+=['\"](" + original_namespace + "[^'\"]*)['\"]"
            match = re.search(pattern, data)
            if match:
                return match.group(1)
            else:
                return original_namespace
        else:
            return original_namespace

    # Need a Timeout someday?
    def call(self, addr, data, namespace, soapaction = None, encoding = None,
        http_proxy = None, config = Config):

        import httplib

        if not isinstance(addr, SOAPAddress):
            addr = SOAPAddress(addr, config)

        # Build a request
        if http_proxy:
            real_addr = http_proxy
            real_path = addr.proto + "://" + addr.host + addr.path
        else:
            real_addr = addr.host
            real_path = addr.path

        if addr.proto == 'httpg':
            from pyGlobus.io import GSIHTTP
            r = GSIHTTP(real_addr, tcpAttr = config.tcpAttr)
        elif addr.proto == 'https':
            r = httplib.HTTPS(real_addr)
        else:
            r = httplib.HTTP(real_addr)

        r.putrequest("POST", real_path)

        r.putheader("Host", addr.host)
        r.putheader("User-agent", SOAPUserAgent())
        t = 'text/xml';
        if encoding != None:
            t += '; charset="%s"' % encoding
        r.putheader("Content-type", t)
        r.putheader("Content-length", str(len(data)))

        # if user is not a user:passwd format
        #    we'll receive a failure from the server. . .I guess (??)
        if addr.user != None:
            val = base64.encodestring(addr.user)
            r.putheader('Authorization','Basic ' + val.replace('\012',''))

        # This fixes sending either "" or "None"
        if soapaction == None or len(soapaction) == 0:
            r.putheader("SOAPAction", "")
        else:
            r.putheader("SOAPAction", '"%s"' % soapaction)

        if config.dumpHeadersOut:
            s = 'Outgoing HTTP headers'
            debugHeader(s)
            print "POST %s %s" % (real_path, r._http_vsn_str)
            print "Host:", addr.host
            print "User-agent: SOAPpy " + __version__ + " (http://pywebsvcs.sf.net)"
            print "Content-type:", t
            print "Content-length:", len(data)
            print 'SOAPAction: "%s"' % soapaction
            debugFooter(s)

        r.endheaders()

        if config.dumpSOAPOut:
            s = 'Outgoing SOAP'
            debugHeader(s)
            print data,
            if data[-1] != '\n':
                print
            debugFooter(s)

        # send the payload
        r.send(data)

        # read response line
        code, msg, headers = r.getreply()

        if headers:
            content_type = headers.get("content-type","text/xml")
            content_length = headers.get("Content-length")
        else:
            content_type=None
            content_length=None

        # work around OC4J bug which does '<len>, <len>' for some reaason
        if content_length:
            comma=content_length.find(',')
            if comma>0:
                content_length = content_length[:comma]

        # attempt to extract integer message size
        try:
            message_len = int(content_length)
        except:
            message_len = -1

        if message_len < 0:
            # Content-Length missing or invalid; just read the whole socket
            # This won't work with HTTP/1.1 chunked encoding
            data = r.getfile().read()
            message_len = len(data)
        else:
            data = r.getfile().read(message_len)

        if(config.debug):
            print "code=",code
            print "msg=", msg
            print "headers=", headers
            print "content-type=", content_type
            print "data=", data

        if config.dumpHeadersIn:
            s = 'Incoming HTTP headers'
            debugHeader(s)
            if headers.headers:
                print "HTTP/1.? %d %s" % (code, msg)
                print "\n".join(map (lambda x: x.strip(), headers.headers))
            else:
                print "HTTP/0.9 %d %s" % (code, msg)
            debugFooter(s)

        def startswith(string, val):
            return string[0:len(val)] == val

        if code == 500 and not \
               ( startswith(content_type, "text/xml") and message_len > 0 ):
            raise HTTPError(code, msg)

        if config.dumpSOAPIn:
            s = 'Incoming SOAP'
            debugHeader(s)
            print data,
            if (len(data)>0) and (data[-1] != '\n'):
                print
            debugFooter(s)

        if code not in (200, 500):
            raise HTTPError(code, msg)


        # get the new namespace
        if namespace is None:
            new_ns = None
        else:
            new_ns = self.getNS(namespace, data)

        # return response payload
        return data, new_ns

################################################################################
# SOAP Proxy
################################################################################
class SOAPProxy:
    def __init__(self, proxy, namespace = None, soapaction = None,
                 header = None, methodattrs = None, transport = HTTPTransport,
                 encoding = 'UTF-8', throw_faults = 1, unwrap_results = None,
                 http_proxy=None, config = Config, noroot = 0,
                 simplify_objects=None):

        # Test the encoding, raising an exception if it's not known
        if encoding != None:
            ''.encode(encoding)

        # get default values for unwrap_results and simplify_objects
        # from config
        if unwrap_results is None:
            self.unwrap_results=config.unwrap_results
        else:
            self.unwrap_results=unwrap_results

        if simplify_objects is None:
            self.simplify_objects=config.simplify_objects
        else:
            self.simplify_objects=simplify_objects

        self.proxy          = SOAPAddress(proxy, config)
        self.namespace      = namespace
        self.soapaction     = soapaction
        self.header         = header
        self.methodattrs    = methodattrs
        self.transport      = transport()
        self.encoding       = encoding
        self.throw_faults   = throw_faults
        self.http_proxy     = http_proxy
        self.config         = config
        self.noroot         = noroot

        # GSI Additions
        if hasattr(config, "channel_mode") and \
               hasattr(config, "delegation_mode"):
            self.channel_mode = config.channel_mode
            self.delegation_mode = config.delegation_mode
        #end GSI Additions

    def invoke(self, method, args):
        return self.__call(method, args, {})

    def __call(self, name, args, kw, ns = None, sa = None, hd = None,
        ma = None):

        ns = ns or self.namespace
        ma = ma or self.methodattrs

        if sa: # Get soapaction
            if type(sa) == TupleType:
                sa = sa[0]
        else:
            if self.soapaction:
                sa = self.soapaction
            else:
                sa = name

        if hd: # Get header
            if type(hd) == TupleType:
                hd = hd[0]
        else:
            hd = self.header

        hd = hd or self.header

        if ma: # Get methodattrs
            if type(ma) == TupleType: ma = ma[0]
        else:
            ma = self.methodattrs
        ma = ma or self.methodattrs

        m = buildSOAP(args = args, kw = kw, method = name, namespace = ns,
            header = hd, methodattrs = ma, encoding = self.encoding,
            config = self.config, noroot = self.noroot)


        call_retry = 0
        try:

            r, self.namespace = self.transport.call(self.proxy, m, ns, sa,
                                                    encoding = self.encoding,
                                                 http_proxy = self.http_proxy,
                                                    config = self.config)

        except Exception, ex:
            #
            # Call failed.
            #
            # See if we have a fault handling vector installed in our
            # config. If we do, invoke it. If it returns a true value,
            # retry the call.
            #
            # In any circumstance other than the fault handler returning
            # true, reraise the exception. This keeps the semantics of this
            # code the same as without the faultHandler code.
            #

            if hasattr(self.config, "faultHandler"):
                if callable(self.config.faultHandler):
                    call_retry = self.config.faultHandler(self.proxy, ex)
                    if not call_retry:
                        raise
                else:
                    raise
            else:
                raise

        if call_retry:
            r, self.namespace = self.transport.call(self.proxy, m, ns, sa,
                                                    encoding = self.encoding,
                                                    http_proxy = self.http_proxy,
                                                    config = self.config)


        p, attrs = parseSOAPRPC(r, attrs = 1)

        try:
            throw_struct = self.throw_faults and \
                isinstance (p, faultType)
        except:
            throw_struct = 0

        if throw_struct:
            if Config.debug:
                print p
            raise p

        # If unwrap_results=1 and there is only element in the struct,
        # SOAPProxy will assume that this element is the result
        # and return it rather than the struct containing it.
        # Otherwise SOAPproxy will return the struct with all the
        # elements as attributes.
        if self.unwrap_results:
            try:
                count = 0
                for i in p.__dict__.keys():
                    if i[0] != "_":  # don't count the private stuff
                        count += 1
                        t = getattr(p, i)
                if count == 1: # Only one piece of data, bubble it up
                    p = t
            except:
                pass

        # Automatically simplfy SOAP complex types into the
        # corresponding python types. (structType --> dict,
        # arrayType --> array, etc.)
        if self.simplify_objects:
            p = simplify(p)

        if self.config.returnAllAttrs:
            return p, attrs
        return p

    def _callWithBody(self, body):
        return self.__call(None, body, {})

    def __getattr__(self, name):  # hook to catch method calls
        if name == '__del__':
            raise AttributeError, name
        return self.__Method(self.__call, name, config = self.config)

    # To handle attribute wierdness
    class __Method:
        # Some magic to bind a SOAP method to an RPC server.
        # Supports "nested" methods (e.g. examples.getStateName) -- concept
        # borrowed from xmlrpc/soaplib -- www.pythonware.com
        # Altered (improved?) to let you inline namespaces on a per call
        # basis ala SOAP::LITE -- www.soaplite.com

        def __init__(self, call, name, ns = None, sa = None, hd = None,
            ma = None, config = Config):

            self.__call 	= call
            self.__name 	= name
            self.__ns   	= ns
            self.__sa   	= sa
            self.__hd   	= hd
            self.__ma           = ma
            self.__config       = config
            return

        def __call__(self, *args, **kw):
            if self.__name[0] == "_":
                if self.__name in ["__repr__","__str__"]:
                    return self.__repr__()
                else:
                    return self.__f_call(*args, **kw)
            else:
                return self.__r_call(*args, **kw)

        def __getattr__(self, name):
            if name == '__del__':
                raise AttributeError, name
            if self.__name[0] == "_":
                # Don't nest method if it is a directive
                return self.__class__(self.__call, name, self.__ns,
                    self.__sa, self.__hd, self.__ma)

            return self.__class__(self.__call, "%s.%s" % (self.__name, name),
                self.__ns, self.__sa, self.__hd, self.__ma)

        def __f_call(self, *args, **kw):
            if self.__name == "_ns": self.__ns = args
            elif self.__name == "_sa": self.__sa = args
            elif self.__name == "_hd": self.__hd = args
            elif self.__name == "_ma": self.__ma = args
            return self

        def __r_call(self, *args, **kw):
            return self.__call(self.__name, args, kw, self.__ns, self.__sa,
                self.__hd, self.__ma)

        def __repr__(self):
            return "<%s at %d>" % (self.__class__, id(self))
